### ISSUE TITLE
Actual static linking / Not linking to glibc

### ISSUE THREAD

**Issue by the-maldridge** [hashicorp/nomad#5643]
Per suggestion in #5537 and encouragement from @angrycub, I'm opening this issue for the purpose of collecting thumbs-ups for building Nomad against non-glibc C libraries.  As the binaries currently provided on https://nomadproject.io/ are not statically linked they only work if the machine they're running on has the correct version of glibc available.

Given that the Go way is to provide binaries that will run anywhere the CPU arch matches, I personally find this surprising.  I find is even more surprising given that the effort to build in a container that contains an alternate C library is minimal.

While not getting into the justification of why one would run Nomad on a non-libC platform as it is beyond the scope of this issue, some quick reasons that jump to mind from my own environments where muslc is the library of choice:

- muslc is a smaller and more correct implementation of the C Standard Library.  Its very difficult to do something so badly in Go that a hard dependency on glibc is introduced, so artificially introducing the dependency by policy is an odd, and currently unjustified in writing, choice.
- Distribution provided packages aren't just for glibc (see the patches that @nilium is currently maintaining for Void Linux's builds)
- Nomad is distributed as a "static" binary, not a distribution package.  Either Hashicorp supports distributions or they don't, but distributing a binary which should run on any Linux distribution, and then claiming that those distributions aren't supported is deceptive.
- muslc is the library of choice for building containers, and a number of distributions are now available that allow the benefits of musl to come to the metal.  Given that the systems that nomad supervises tend to be minimal, it seems an obvious choice to pick a minimal system to run on the metal, thus maximizing Mhz and RAM available to tasks.

If builds without an artificial dependency on glibc is something that is important to you, please thumbsup this issue (or even just the support to create these builds on your own, as you should be able to do already but can't with the current release).  Don't reply with a +1 as that just clutters the thread and makes your support difficult to track.

**Comment by schmichael** [comment]
Sorry for the late reply. This is a complex issue that defies a quick and risk-free fix. I'll try to respond to each of your points:

> Given that the Go way is to provide binaries that will run anywhere the CPU arch matches, I personally find this surprising.

While Go is renowned for producing binaries that can run on a wide variety of systems, it's far from a perfect solution. The Go toolchain distributed by golang.org does not run on Alpine. [Alpine ships their own customized build of the Go toolchain and applies patches for compatibility with musl.](https://git.alpinelinux.org/aports/tree/community/go?h=master)

Nomad requires CGO (for [libcontainer](https://github.com/opencontainers/runc/tree/master/libcontainer/nsenter) among other things) which complicates portability.

> Its very difficult to do something so badly in Go that a hard dependency on glibc is introduced, so artificially introducing the dependency by policy is an odd, and currently unjustified in writing, choice.

Our Nvidia device driver implementation requires glibc due to differences in [glibc's lazy binding.](https://wiki.musl-libc.org/functional-differences-from-glibc.html#Lazy-bindings)

> Nomad is distributed as a "static" binary, not a distribution package.

If we shipped _all_ of Nomad's plugins as external binaries, the agent itself could probably run on musl (perhaps even drop cgo entirely). This would require significant changes to our source layout, build system, distributed artifacts, documentation, and test infrastructure. We have no plans to produce "lite" portable agent binaries at this time.

> If builds without an artificial dependency on glibc is something that is important to you

We would love more assistance in improving Nomad's portability and Alpine/musl support. I hope we have demonstrated in #5537 that we do respond to specific issues and put effort toward improving our portability story.

As far as I know making the prebuilt Nomad binaries work on Alpine/musl does not have a straightforward solution, but I'd love to be wrong or have assistance in working toward portability.

**Comment by the-maldridge** [comment]
@schmichael no worries at all on the slow reply, honestly I didn't expect to ever get a reply on this ticket.

As a Void maintainer, I'm well aware of the hoops that are needed to get a full blown toolchain up and running, however I'm also not aware of any projects out there that have such deep dependencies (well maybe glibc, but that's another point entirely) that they care about the toolchain they're built with.

I live in hope that the libcontainer dependency will move out into a plugin.  As that system becomes more robust, I'd really like to get to a point where nomad itself is running rootless.

I'm well acquainted with glibc's generally broken lazy binding system and the sad state of affairs that is GPU driver options on Linux.  I'm very happy that there was a quick solution in adding the build tag to shut it off, though it would have been nice to get that into an actual release, rather than delaying it for a while since in the official binaries it was a noop change.

I'd actually really like to see nomad shipping more plugins as external binaries that could either be setgid or setuid as necessary.  Running the whole nomad binary as root is a really interesting security story that I'm still trying to grapple with.  I can see why this is nowhere on the roadmap, as its a complete overhaul of a lot of the core parts of Nomad; as a means of improving the amount of code running with elevated privileges, I really hope this approach can be considered.

#5537 was a fantastic start that was unfortunately marred by the fix not making it into the release.  What it demonstrated from my perspective was that getting the fixes into the codebase was something that Hashicorp was willing to do, but putting them into another release - and unbreaking a class of users in the process - was another.

The solution to running Nomad on Alpine is to compile from source.  The nvidia integration has to be switched off to make it work, but as GPUs tend to be a specialized accelerator that's not on the vast vast majority of the fleet, this is a level of complexity I'm willing to pay.

Useful takeway from this ticket: the single biggest thing that could be useful right now to supporting Nomad on non-glibc systems would be a smoke-test build before every release.  This is making sure that the build works at all, as there have now been 2 releases where `nomad version` did not work after a build.  Of course this smoke test build only has value if it can block a release, but small steps and all that.

**Comment by prologic** [comment]
(_not tested myself_) Can the Nomad binaries be compiled and linked statically today with the current source tree and layout?

**Comment by the-maldridge** [comment]
Yes, you can.  Its very hard to build something that *can't* be statically linked.  Here's some commands that will do it for you in Alpine:

```
    apk add \
        bash \
        g++ \
        git \
        linux-headers \
        musl-dev

    # Hashicorp Build
    mkdir -p src/github.com/hashicorp/nomad
    cd src/github.com/hashicorp/nomad || return 1

    # Get source and apply any patches.
    git clone -b v$version https://github.com/hashicorp/nomad.git .

    echo "Building..."
    go build -x \
       -o bin/nomad \
       -tags "nonvidia release ui" \
       --ldflags '-linkmode external -extldflags "-static"' \
       .
```

That's yanked from some obsolete files from when I was linking statically, now I maintain an internal Alpine repo which has the binaries dynamically linked to musl.

**Comment by prologic** [comment]
The `-tags "nonvidia` bit is probably the important one that drags in cgo :)

**Comment by the-maldridge** [comment]
Not quite.  cgo is pulled in by the DNS resolver.  This is a fairly well understood gotcha at this point.  The nvidia issue is that the bindings for that driver perform unchecked operations with no error handling, so if you don't have the shared objects available then you're SOL, since musl doesn't provide the same lazy bindings that glibc does (hence why this works on glibc if you don't have the soname, but not on musl).

**Comment by the-maldridge** [comment]
After thinking long and hard about this issue, I'm not sure in its presents state traction can be made.  Too many of Nomad's drivers need to link system level code.  I think the only way to make progress on this would be to factor out all the drivers from Nomad into go-plugin executables.  As a practical upshot and something that might be easier to sell to the product owners, this would mean the main, *network exposed* nomad binary doesn't need to be running with euid 0.

I'm not sure what the overall desire within HashiCorp for this change would be, but it would be a pretty big win for security since increasingly less code would need to run in a privileged mode.  If the podman driver that's been talked about on gitter ever gets off the ground, it would be trivial to run completely rootless nomad, and that would make my security team very happy indeed.

**Comment by schmichael** [comment]
> After thinking long and hard about this issue, I'm not sure in its present state traction can be made. ... I think the only way to make progress on this would be to factor out all the drivers from Nomad into go-plugin executables.

:+1:  I'd love to ship a minimal Nomad binary (all plugins external) _alongside our monolithic binary_, but it's probably not going to be prioritized soon (not 0.11 or probably not 0.12) since it's nontrivial effort to support this in our build system.

There are plans underway to ship proper Linux packages alongside our zip files which may give us a route to drop our monolithic binary while still providing a "batteries included" package. That could make this effort substantially easier.

**Comment by the-maldridge** [comment]
That's a really cool idea.  I use go-plugin in some of my own projects, but haven't yet noticed any machinery that would let me selectively bake in plugins to a monolith.  Can you share any pointers to documentation where I can find how that works?

**Comment by schmichael** [comment]
Unfortunately `go-plugin` itself doesn't have a notion of internal or builtin plugins. Builtin plugins define the same interface as external plugins and register themselves manually with the agent's plugin manager: https://github.com/hashicorp/nomad/blob/v0.10.2/helper/pluginutils/catalog/register.go#L11-L20

The PluginLoader [discovers external plugins and merges](https://github.com/hashicorp/nomad/blob/v0.10.2/helper/pluginutils/loader/init.go#L51-L84) them [into the plugin catalog.](https://pkg.go.dev/github.com/hashicorp/nomad@v0.10.2/helper/pluginutils/loader?tab=doc#PluginLoader.Catalog)

The [PluginLoader's Dispense method](https://pkg.go.dev/github.com/hashicorp/nomad@v0.10.2/helper/pluginutils/loader?tab=doc#PluginLoader.Dispense) is kind of the bridge between internal and external plugins as it executes the binary for external plugins.

Just brainstorming but an approach to migrating to shipping monolithic and minimal builds would be to create hidden Nomad subcommands for executing builtin plugins: eg `nomad _plugin-exec <name>` would start the go-plugin subprocess within the monolithic Nomad binary. A registration pattern could be used to statically build a list of builtin plugins at compile-time (and therefore be empty for the minimal build).

**Comment by the-maldridge** [comment]
Hmm, I like the idea of the hidden subcommands.  It seems really clean from an interfaces perspective, if not an implementation one.

**Comment by brodul** [comment]
![image](https://user-images.githubusercontent.com/204514/92112200-b9e44b80-eded-11ea-81a0-6e46abe5bc5b.png)

I use a weird distribution of Linux that does't have interpreter, bash or libs on standard location.
The community uses a program called `patchelf` to modify ELF headers of close source programs. https://github.com/NixOS/patchelf 
In this case it can be used as a workaround.

![image](https://user-images.githubusercontent.com/204514/92112568-455ddc80-edee-11ea-8deb-05b230e8271d.png)

**Comment by the-maldridge** [comment]
Interesting use of patchelf, I was unaware it was safe to use on Go compiled binaries.  My use case has switched to running Nomad in a chroot that has the loader and libraries that Nomad expects to see, but I do still compile out the nvidia extensions as they are just more trouble than they're worth.

**Comment by tgross** [comment]
While I don't think it entirely resolves this issue because we still have `libcontainer` in the mix, https://github.com/hashicorp/nomad/pull/10796 externalizes the nvidia device plugin (finally).

**Comment by jm-positron** [comment]
It seems that plugin creation also depends on cgo (though musl can be used in that case).

It would be helpful to understand why that is the case, and if that dependency could be removed before needing to tackle the bigger issue in the primary nomad binary.

It seems related to this discussion, but let me know if I should open a new issue about this.

**Comment by tgross** [comment]
Hey @jm-positron writing a driver shouldn't require CGO. That may be an issue with the plugin SDK (such as it is). Can you open a new issue describing what you're hitting with some example code?

**Comment by schmichael** [comment]
_Actual_ static linking is available as a **custom** build target thanks to https://github.com/hashicorp/nomad/pull/27310

To produce a statically linked binary run:

```
make dev-static

# or for a full release with codegen'd codecs
make release-static
```

Note that this required disabling the `exec` and `java` drivers in the build, so we can't make this the default release.

Testing and feedback welcome!

The [`exec2`](https://github.com/hashicorp/nomad-driver-exec2) is hopefully preferable to `exec` in every way, and we could externalize `exec` and `java` into plugins like we did with [`rkt`](https://github.com/hashicorp/nomad-driver-rkt) a long time ago (RIP).

Unlike `rkt`, `exec` has proven fairly popular, so I'm not sure how quickly we could move on its removal.

**Comment by jm-positron** [comment]
> Hey [@jm-positron](https://github.com/jm-positron) writing a driver shouldn't require CGO. That may be an issue with the plugin SDK (such as it is). Can you open a new issue describing what you're hitting with some example code?

To close the loop on this, for anyone reading it, I believe drivers _do not_ require CGO. My previous comment was the result of a build command error. We are able to successfully build a driver without CGO. Apologies for the incorrect statement.

**Comment by the-maldridge** [comment]
@schmichael I think you're facing a similar problem to a lot of the other hashicorp products, which is that navigating the documentation is an exercise in frustration.  I'd like to consider myself fairly knowledgeable about Nomad and its ecosystem, but I could not effectively tell you all the differences in exec and exec2, or about some of the other external plugins and what the tradeoffs are.  Examples that come to mind are not knowing the libvirt plugin was in a usable state until running into a Nomad PM at HashiConf, the status of the containerd plugin still being "who maintains this?" and the general questions around release compatibility as different plugins don't have releases in lock step with Nomad itself, so its unclear what is and isn't safe to update at any given time.  Reading the documentation really leaves a taste of "here's nomad and it does all this cool stuff, and by the way here are some plugins" and not much more information, which makes it difficult to make an informed choice about whether or not a plugin is a better choice than a builtin.
